File: test/validator-auth-p0/codec.cpp
// Independent ordered-schema decoder and encoder for interoperability checks.
#include <cstdint>
#include <fstream>
#include <iostream>
#include <iterator>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>
struct Type {std::string tag;std::vector<std::string> fields;};
std::map<std::string,Type> types;
std::vector<std::string> split(const std::string& s,char c){std::vector<std::string> out;std::stringstream in(s);std::string p;while(std::getline(in,p,c))out.push_back(p);return out;}
struct Codec {
 std::vector<unsigned char> input,output;std::size_t at=0;
 void require(bool b){if(!b)throw std::runtime_error("refused");}
 std::uint64_t read(unsigned n){require(n<=input.size()-at);std::uint64_t v=0;for(unsigned i=0;i<n;++i)v=(v<<8)|input[at++];return v;}
 void write(std::uint64_t v,unsigned n){for(unsigned i=n;i>0;--i)output.push_back(static_cast<unsigned char>(v>>(8*(i-1))));}
 void bytes(std::size_t n){require(n<=input.size()-at);output.insert(output.end(),input.begin()+at,input.begin()+at+n);at+=n;}
 void value(const std::string& t,unsigned depth=0){
  require(depth<64);
  if(t=="u8"||t=="u16"||t=="u32"||t=="u64"||t=="i32"){unsigned n=std::stoul(t.substr(1))/8;write(read(n),n);}
  else if(t=="h")bytes(32);
  else if(t[0]=='b'){auto n=read(4);require(n<=std::stoull(t.substr(1)));write(n,4);bytes(n);}
  else if(t[0]=='l'){auto f=split(t.substr(1),'/');require(f.size()==3);unsigned w=std::stoul(f[0])/8;auto n=read(w);require(n<=std::stoull(f[1]));write(n,w);for(std::uint64_t i=0;i<n;++i)value(f[2],depth+1);}
  else {auto found=types.find(t);require(found!=types.end());const auto& def=found->second;
   if(def.tag!="-"){for(unsigned char c:def.tag){require(read(1)==c);write(c,1);}require(read(2)==1);write(1,2);require(read(2)==0);write(0,2);}
   for(const auto& child:def.fields)value(child,depth+1);
  }
  require(output.size()<=33554432);
 }
};
int main(int argc,char** argv){try{
 if(argc!=5)throw std::runtime_error("arguments");
 std::ifstream schema(argv[1]);if(!schema)throw std::runtime_error("schema");std::string line;
 while(std::getline(schema,line)){auto f=split(line,'|');if(f.size()<2)throw std::runtime_error("schema line");types.emplace(f[0],Type{f[1],f.size()>2?split(f[2],','):std::vector<std::string>{}});}
 std::ifstream in(argv[3],std::ios::binary);if(!in)throw std::runtime_error("input");Codec c;c.input.assign(std::istreambuf_iterator<char>(in),{});c.require(c.input.size()<=33554432);c.value(argv[2]);c.require(c.at==c.input.size());
 std::ofstream out(argv[4],std::ios::binary);out.write(reinterpret_cast<const char*>(c.output.data()),c.output.size());if(!out)throw std::runtime_error("output");
 }catch(const std::exception& e){std::cerr<<e.what()<<'\n';return 1;}}

File: test/validator-auth-p0/codec.rs
// Independent ordered-schema decoder and encoder for interoperability checks.
use std::{collections::BTreeMap,env,fs};
#[derive(Clone)]
struct Type {tag:String,fields:Vec<String>}
struct Codec {input:Vec<u8>,output:Vec<u8>,at:usize,types:BTreeMap<String,Type>}
impl Codec {
 fn read(&mut self,n:usize)->Result<u64,String>{
  if n>self.input.len()-self.at{return Err("truncated".into());}
  let mut v=0;for _ in 0..n{v=(v<<8)|u64::from(self.input[self.at]);self.at+=1;}Ok(v)
 }
 fn write(&mut self,v:u64,n:usize){for i in (0..n).rev(){self.output.push((v>>(8*i)) as u8);}}
 fn bytes(&mut self,n:usize)->Result<(),String>{
  if n>self.input.len()-self.at{return Err("truncated".into());}
  self.output.extend_from_slice(&self.input[self.at..self.at+n]);self.at+=n;Ok(())
 }
 fn number(t:&str)->Result<usize,String>{t.parse::<usize>().map_err(|e|e.to_string())}
 fn value(&mut self,t:&str,depth:usize)->Result<(),String>{
  if depth>=64{return Err("depth".into());}
  if ["u8","u16","u32","u64","i32"].contains(&t){let n=Self::number(&t[1..])?/8;let v=self.read(n)?;self.write(v,n);}
  else if t=="h"{self.bytes(32)?;}
  else if let Some(b)=t.strip_prefix('b'){let n=self.read(4)? as usize;if n>Self::number(b)?{return Err("blob-bound".into());}self.write(n as u64,4);self.bytes(n)?;}
  else if let Some(l)=t.strip_prefix('l'){
   let f:Vec<_>=l.split('/').collect();if f.len()!=3{return Err("schema".into());}
   let w=Self::number(f[0])?/8;let n=self.read(w)? as usize;if n>Self::number(f[1])?{return Err("list-bound".into());}
   self.write(n as u64,w);for _ in 0..n{self.value(f[2],depth+1)?;}
  }else{
   let def=self.types.get(t).ok_or("unknown type")?.clone();
   if def.tag!="-"{for c in def.tag.bytes(){if self.read(1)?!=u64::from(c){return Err("tag".into());}self.write(u64::from(c),1);}
    if self.read(2)?!=1{return Err("version".into());}self.write(1,2);
    if self.read(2)?!=0{return Err("flags".into());}self.write(0,2);
   }
   for child in def.fields{self.value(&child,depth+1)?;}
  }
  if self.output.len()>33554432{return Err("object-bound".into());}Ok(())
 }
}
fn run()->Result<(),String>{
 let args:Vec<_>=env::args().collect();if args.len()!=5{return Err("arguments".into());}
 let schema=fs::read_to_string(&args[1]).map_err(|e|e.to_string())?;let mut types=BTreeMap::new();
 for line in schema.lines(){let f:Vec<_>=line.split('|').collect();if f.len()!=3{return Err("schema".into());}
  let fields=if f[2].is_empty(){vec![]}else{f[2].split(',').map(String::from).collect()};types.insert(f[0].into(),Type{tag:f[1].into(),fields});
 }
 let input=fs::read(&args[3]).map_err(|e|e.to_string())?;if input.len()>33554432{return Err("object-bound".into());}
 let mut c=Codec{input,output:vec![],at:0,types};c.value(&args[2],0)?;
 if c.at!=c.input.len(){return Err("trailing".into());}fs::write(&args[4],c.output).map_err(|e|e.to_string())
}
fn main(){if let Err(e)=run(){eprintln!("{}",e);std::process::exit(1);}}
